output/helpers_test.go

```diff
@@ -0,0 +1,118 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package output
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/loadimpact/k6/stats"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSampleBufferBasics(t *testing.T) {
+	single := stats.Sample{
+		Time:   time.Now(),
+		Metric: stats.New("my_metric", stats.Rate),
+		Value:  float64(123),
+		Tags:   stats.NewSampleTags(map[string]string{"tag1": "val1"}),
+	}
+	connected := stats.ConnectedSamples{Samples: []stats.Sample{single, single}, Time: single.Time}
+	buffer := SampleBuffer{}
+
+	assert.Empty(t, buffer.GetBufferedSamples())
+	buffer.AddMetricSamples([]stats.SampleContainer{single, single})
+	buffer.AddMetricSamples([]stats.SampleContainer{single, connected, single})
+	assert.Equal(t, []stats.SampleContainer{single, single, single, connected, single}, buffer.GetBufferedSamples())
+	assert.Empty(t, buffer.GetBufferedSamples())
+
+	// Verify some internals
+	assert.Equal(t, cap(buffer.buffer), 5)
+	buffer.AddMetricSamples([]stats.SampleContainer{single, connected})
+	buffer.AddMetricSamples(nil)
+	buffer.AddMetricSamples([]stats.SampleContainer{})
+	buffer.AddMetricSamples([]stats.SampleContainer{single})
+	assert.Equal(t, []stats.SampleContainer{single, connected, single}, buffer.GetBufferedSamples())
+	assert.Equal(t, cap(buffer.buffer), 4)
+	buffer.AddMetricSamples([]stats.SampleContainer{single})
+	assert.Equal(t, []stats.SampleContainer{single}, buffer.GetBufferedSamples())
+	assert.Equal(t, cap(buffer.buffer), 3)
+	assert.Empty(t, buffer.GetBufferedSamples())
+}
+
+//nolint:gosec
+func TestSampleBufferConcurrently(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
```

Comment:
From the perspective that this might get other tests and if one of the random values was the actual reason behind this failed I would prefer if:
1. we don't use the global but a local instance
2. we t.Log the random seed
I think that you can just also generate a chan int64 with 
```golang
var signal :+ make(chan struct{})
defer close(signal)
var ch := make(chan int64, 50)
go func() {
for {
   select {
   case ch<- rand.Int63n(1000):
   case <- signal:
     return
   }
}()
```
So you don't have to wrap the rand.Source  so it is concurrently safe.

(Code above was written in this text fields so it will probably not work :P )

Reply:
I'll remove rand altogether, it doesn't really bring any benefit... In fact, the minimum synchronization it requires (be it a channel or the stdlib's locks) might prevent us from having a concurrency issue surface, which is the point of the test.

Reply:
Went half-way, made the randomness predictable and took it out of the goroutines: https://github.com/loadimpact/k6/pull/1869/commits/cbe173bdefbe64241a64826d8c14c7c8297cca9c